### PR TITLE
add option to sync after every write

### PR DESCRIPTION
**Pull request by adammoody**
This adds an option to sync after every (successful) write operation.  This will slow down performance, but it could be a work around for codes that are missing the necessary sync calls.

This includes a runtime option to enable the auto-sync-after-write ``UNIFYFS_CLIENT_WRITE_SYNC=1``.

It should help with https://github.com/LLNL/UnifyFS/issues/560

### CODE REVIEW

**Comment by adammoody** [comment]
@clmendes, I haven't tested this at all other than testing that it compiles, but I think the MPI I/O test code should work with this, and hopefully the HDF example.  If it helps, I could rebase this on the margotree file hash branch.

**Comment by clmendes** [comment]
@adammoody , thanks for providing this. I can test it with the MPI-IO reproducer that I had created. However, for the "real" tests (i.e. tests from the HDF5 testsuite), I'd need to use it with at least the "margotree" branch (or after PR-566 is merged), since some of the other HDF5 tests only worked after I started to use the margotree branch from Eric.

**Comment by adammoody** [comment]
@clmendes , oh it looks like it doesn't quite apply cleanly to the margotree branch proper (too old I guess).

I think I could apply it to the branch that is hosting PR #566 (new-margotree-abtsync-filehash).  Have you moved on to testing with branch yet?  If so, I'll target that.

**Comment by clmendes** [comment]
@adammoody , no, I have not yet tested the PR-566's branch. Today I had a good part of the day around SC20 (we had our Blue Waters talk presented today, among other things), so not very productive.

But since this sync change could be applied to the mpi-io test separately, I was planning to do it using your 'writesync' branch, which I obtained yesterday. This would be quicker to test than the rest of the HDF5 tests (for those I'm indeed counting on that branch in PR-566).

**Comment by clmendes** [comment]
@adammoody , I finally tested this new branch with my reproducer code from PR-560 (the mpi-io example), sorry for my delay. I have good and bad news from today's test.

The good news is that by setting UNIFYFS_CLIENT_WRITE_SYNC=1 in my job script the program indeed reads 12 bytes as expected. I even changed the example to print those 12 bytes, and verified that they had the correct value.

Now, the bad news is that even with this, the resulting file (datafile-u) is different from the file that I obtain without Unify (datafile-m), and the difference lies in positions [101-107]. In other words, that last write done by rank-0, of 12 bytes starting at offset=95, does not make completely to the resulting file: only the first 5 bytes are actually present in the resulting file. Bytes [101-107] still have the original value, initially written by rank-1. To fix this, I need to use the 'margotree' branch! (as a reminder, using the margotree branch, *and* adding the MPI_File_sync before the first barrier, everything works fine, i.e. 12 bytes are read and the resulting file is exactly the same as the file produced by the non-Unify version).

So, in summary, I believe your change will be valid, but to use it I'd really need to combine it with the 'margotree' branch, or with that newer branch from PR-566! As it stands right now, it fixes this write/sync issue, but there are other Unify problems that only got fixed after the margotree branch was created!

**Comment by adammoody** [comment]
This writesync branch has now been rebased on dev since merging the margotree code.

**Review by MichaelBrim — APPROVED** [review]
Looks good to me.

**Comment by adammoody** [comment]
Thanks, @MichaelBrim

**Comment by adammoody** [comment]
@clmendes , this has now been merged into dev, so a fresh pull of dev should enable that option for you now.
